### ISSUE TITLE
Parallelism control per-resource when count > 1

### ISSUE THREAD

**Issue by mengesb** [hashicorp/terraform#7631]
Feature/Enhancement request

I'd like to see support for modifying the parallelism on a per-resource basis.

Use cases:
1. Create a quantity of the same resource, but actions must be executed one at a time. An example of this would be chef-backend, and joining the cluster. A leader (the first) must be established, but then subsequent backend members (initially followers) must join one at a time due to the backend configuration steps going on in the background. This would avoid the need to craft a waiting mechanism for N objects (here's where `count` comes into play)

2. For load considerations, sometimes more or less than the current parallelisim can be handled. It would be nice to have control per resource on this.

3. In the case of a cluster initiator and cluster members joining (sequential) we have the possibility to control initialization versus joining, and do so in sequence in one resource versus two or more with complicated logic.

4. Sequential phases are more tightly controlled, and doesn't affect `depends_on` but reduces overall line count and appearances of resource duplication

**Comment by matthughes** [comment]
Would love this as well.  Above use cases ring true.  I also have an issue this would solve with OpenStack server_group / anti-affinity.

Scheduling 10 instances with anti-affinity in parallel blows up pretty reliably on our Openstack deployment.  It seems like the server maintains a global lock and only allows one instance to be scheduled at a time, leading to timeouts.  There is another issue floating around here for custom timeout per resource, but I think this is a more elegant solution for most cases.

**Comment by dfairhurst** [comment]
I'd like to see this.  Our Openstack deployment can only handle adding security rules to a group one at a time.  It takes all the requests, but processes them sequentially.  The problem is that even if Terraform sends 100 requests in parallel (-parallelism 100) they'll only be processed one by one.  If I could specify through the resource that the rules are only to be done with parallelism 1, then Terraform would be able to do other operations in parallel instead and optimise the deployment time.

**Comment by herasimau** [comment]
Any news about this feature?

**Comment by MaxFlanders** [comment]
+1 this would be very helpful, am building a clustered system where masters are identical (can be built using the same resource), but need to be joined to the cluster 1 at a time.

**Comment by apparentlymart** [comment]
Hi all,

For the moment we don't have any specific plans in this area, due to attention being elsewhere. From re-reading the discussion here I see two main use-cases:

* Backend cannot deal with so many requests at the same time: in the ideal world, this isn't something a Terraform user should be worrying about, and instead the provider itself should be insulating Terraform from limitations of the underlying API to make it work within Terraform's model. At the moment the provider SDK lacks any first-class features to help with this, and so I expect we will look for ways to improve on this in future SDK changes, but in the mean time some providers use a whole-provider-process synchronization primitive (e.g. a semaphore) to limit concurrency.

* The ordering of the creation of the instances needs to be controlled in some way, such as if one of them is "special". We intend `count` to be for creating multiple instances that are _equivalent_, so this use-case is slightly outside of its scope. In this case I would lean towards being explicit about it and having the special node be a separate resource, and thus the relationship can be seen clearly in configuration:

  ```hcl
  resource "aws_instance" "leader" {
    # ...
  }
  resource "aws_instance" "followers" {
    count = 3

    # ...
    user_data = "leader=${aws_instance.leader.private_ip}"
  }  
  ```

I expect we will re-visit these use-cases in the long run and look for ways to accommodate them better. #7388 is somewhat-related to the first use-case too, since it would allow some providers to reduce load on their backend APIs by coalescing similar operations together.

For the moment, and for the foreseeable future, our attention is unfortunately focused elsewhere.

**Comment by encron** [comment]
Hey @apparentlymart , 

We have another use case where we want to be able to perform a zero-downtime rolling update of clustered stateful services where we persist data on EBS volumes:

The EBS volume has to be detached and attached to a new instance in a sequential order for each instance of that cluster. Since we want to assure there is no downtime, we will need to cycle every instance 1-by-1 while keeping the cluster itself healthy. However, terraform will instead perform create/destroy operations on all instances at the same time (although limited by the global flag `-parallelism`). 

So even with `create_before_destroy = true` it means it will create `n` new instances, then destroy the old ones, yet all at the same time. This will resort in downtime for the gap in which the EBS volumes are being swapped from the old to the new instances + the time it takes for the service to come back up afterwards. This would of course not be an issue if we didn't have to swap the EBS volumes because then we could mark the new instance as "created" only after it joined our cluster and passed health checks while keeping the old one still alive so we'd always have at least 1 healthy instance running.

Therefore, to fix this issue, ideally, we'd like a way to set some sort of mutex/lock where we can:
1. acquire lock
2. create the first new instance of the resource
3. destroy the old one (and detach the EBS volume)
4. attach the EBS volume to the new one
5. perform health checks / wait for it to join the cluster
6. release lock, go to the next instance and repeat 1-5.

You could argue this can be worked around by setting the global `-parallelism` flag to `1` but we don't want to run the whole `terraform apply` with said parallelism nor let potential downtime be depending on a CLI flag that could easily be forgotten.

**Comment by dthvt** [comment]
I'll add another use case:

We are deploying a set of servers behind a load balancer.  Each server automatically uses LetsEncrypt w/ DNS validation to get certificates (via user_data startup scripting).  However, since we don't want to do TLS termination on the load balancer, each server should have a certificate that is valid for "service.domain.com" in addition to "serverX.domain.com".  This is supported by LetsEncrypt, however the DNS validation must thus be done sequentially as the servers as deployed.  As it stands now, deploying the identifical servers using "count = X" results in a race condition where not all servers can successfully validate w/ LetsEncrypt.

**Comment by OffColour** [comment]
This would be extremely useful. It's OK to say it's designed for where they're all identical, but that's the end state. Changing, say, the size of VMs created with count would result in them all rebooting at the same time.
As it stands, I can't use count at all.

**Comment by james-powis** [comment]
I have a reservation system for some physical hardware which is allocated and configured on-demand using models in Terraform, unfortunately the api that finds and reserves the hardware has a pretty nasty race condition which is not easy to overcome.

Disabling parallelism with `-parallelism=1` solves the problem, but should not be an absolute requirement.
